### DISCUSSION TITLE
[@types/aws-lambda] handler not typed correctly for promise usage

### DISCUSSION THREAD

**Discussion by garfieldnate** [DefinitelyTyped/DefinitelyTyped#63474] [Issues with a @types package]
I was having trouble correctly writing aws-lambda handlers, and then I realized that the types are probably wrong. According to the documentation, handlers in aws-lambda can EITHER take a callback OR return a Promise. Therefore, I think the type [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/aws-lambda/handler.d.ts#L84) would more correctly be written as an overload:

     export type Handler<TEvent = any, TResult = any> = ((
      event: TEvent,
      context: Context,
      callback: Callback<TResult>,
    ) => void) | ((
      event: TEvent,
      context: Context
    ) => Promise<TResult>);

The only issue I have with this is that now the parameters for implementing functions get flagged implicitly having `any` type:

    export const myHandler: Handler = async (data, context) => {...}

>Parameter 'data' implicitly has an 'any' type.
>Parameter 'context' implicitly has an 'any' type.

[a comment by an automated account is not shown]

**Comment by joeykilpatrick** [comment]
It's great to have people bring things like this up when they think the types in this library are wrong, but I think it may be time that we add a comment in the code to help others find the rationale for why we haven't accepted this change in the past. Many well-intentioned developers have raised issues or submitted PRs for this change, but they continue to be rejected. Here's why:

With the current Handler type,
```typescript
// Current implementation of Handler
type Handler<TEvent = any, TResult = any> = 
    | ((event: TEvent, context: Context, callback: Callback<TResult>) => void | Promise<TResult>)
    ;

// All of these are valid implementations of Handler
const handler1: Handler = (event, context, callback) => { return; }
const handler2: Handler = (event, context) => { return; }
const handler3: Handler = (event) => { return; }
const handler4: Handler = async (event, context, callback) => { return; }
const handler5: Handler = async (event, context) => { return; }
const handler6: Handler = async (event) => { return; }
```

That `handler4` is the use case you are trying to prevent. You don't want users to be able to return a `Promise` AND to use the provided `callback`. However, the proposed solution does not prevent this:

```typescript
// Proposed implementation of Handler
type NewHandler<TEvent = any, TResult = any> =  
    | ((event: TEvent, context: Context, callback: Callback<TResult>) => void)
    | ((event: TEvent, context: Context) => Promise<TResult>)
    ;

// All of these are still valid implementations of NewHandler
const newHandler1: NewHandler = (event, context, callback) => { return; }
const newHandler2: NewHandler = (event, context) => { return; }
const newHandler3: NewHandler = (event) => { return; }
const newHandler4: NewHandler = async (event, context, callback) => { return; } // This is the one you are trying to prevent, but it still works
const newHandler5: NewHandler = async (event, context) => { return; }
const newHandler6: NewHandler = async (event) => { return; }
```

The reason why it doesn't prevent it is because a value of type `(...) => Promise<unknown>` is in fact assignable to `(...) => void`. More on this topic [at this proposal](https://github.com/microsoft/TypeScript/issues/42709). You can try:
```typescript
const unintuitive: () => void = async () => { return 1; };
```
To make matters far worse, the proposed `NewHandler` allows handlers that throw runtime exceptions when invoked by AWS Lambda.

```typescript
// Old one correctly does not allow this. The param 'unexpected' will not be this object or undefined, it will be the callback passed in by AWS Lambda.
const handler7: Handler = async (event, context, unexpected?: {prop1: {prop2: unknown}}) => { // TS-Error: Type 'Callback<any>' is not assignable to type '[]'.
    if (unexpected && unexpected.prop1.prop2) {
        // ...
    }
    // ...
}

// New one incorrectly allows this. This will cause a runtime error when invoked by AWS Lambda.
const newHandler7: NewHandler = async (event, context, unexpected?: {prop1: {prop2: unknown}}) => { 
    if (unexpected && unexpected.prop1.prop2) {
        // ...
    }
    // ...
}
```
For those who are looking to mock their own async handlers, they should have to provided the same arguments that AWS Lambda will provide. It is easy to add a dummy function when testing.

```typescript
function runMyHandler(handler: Handler) {

    const event = {} as never;
    const context = {} as never;

    // People sometimes expect this to work, but it doesn't
    handler(event, context); // TS-Error: An argument for 'callback' was not provided.
    
    // This is a minor change, but correctly mocks what AWS Lambda will provide
    handler(event, context, () => {});

}
```
[TS-Playground link with all of these code samples.](https://www.typescriptlang.org/play?noImplicitAny=false#code/JYOwLgpgTgZghgYwgAgMIHtwQB5mQbwChkTkB6M5AZXQFsUAHKdB6MYCAZ0IF9DCwAT1Zo4AGzEAjRAGsAPABUASlwCuYvAF5kcEIIB8ybQApozKAH4AXMgCiUc8gA+yTmCigA5s+Qh1YgBpkKDUNa2RlULAASiNDADd0YAATAG5+CjRVBwhwZGBaBjEIenA4dkxkdBhkAAldZOKoAWEUepBG6EVbeNytHT0gyM51ft0DI2RiUhdjU17wGwUevqCETEhcGwwsXDXxKVltg+kEeWHR-VjNBKTknwAFZlpgTghFFRGNK+mSdMJMgBBCRVGpgAAWXBQcBCyHi4hS+UKxVKYHKwEwnFBdQaTUI6xAbmQ4Nx0AAjDZ2p0oJN5qtkATNmB9hJTjJroZ8MEIGBsiBUsg+ASiSSOk0AEyU0k0kwQBbMhkbHAxOIEbm8qD8wX4zF4UXUgDMUrF0Fpcr6HLVIQ1WqFuuJ0oALMbqZM4JxBCAEMg6eA1kq9gyTrJLVzrXyBXbCXrpQBWF1NN0er0+81+xW7FU3K08iPa4Uxk1QABsCdN2ndnu9vqznPVeb4GUoTxY6De9wKRRKfXRlWqOKLLREADkIAB3KlNbryt2DCKfUazibaKakHxzNNgJYrdOM5XbAMKhDBs7HVmyD5RK6qxIpaK-dc17fy-2Zg+Zy0tl5vS9fMA-Nd-iBEF+whKEdFhNxgBBeExERTsUR7CpCWxUcJ2lHVo18cdJ3JGw0NwmVUxfDMmRZQ4zlDetNUjTCiRAHDpUlZACOlM0SL3XAqPDGj83tBj0KLI0WMYot2ItVUw1zXio3o0TqWdETBNdCtk2rTdXzIoNz0oyTqNtchKAUcFXnyLEwKqBjkEEdBVAglB3EELxkDAdBkCYDTkEkVQ8GAPAoJBMd0CgGRuALbDlKaeMlMIpMq2I+lONrHMbVo8KBMI0sYrY1T4prbjpNtfhwtUEBQF5PzgF6GxjEtW97lylNar0niQGQMlIyAygAHkxHuTAUHWHIEDAMRBGQZJ0C4Xx0DwA50DHFyTM4AA6CJIXcmE4FoZAAHJSpwVgRogZJduQMdoLEGa8EkBzlqqSQACsIBGqoaVK5IIBgUATqCPzzsury7sGk8ZE2zh23yNrJHGwEAHUqGQAAZbbJGSOAVrowtqQAdjLIjKyazykqCA7sCOyBknCfAmBYCkCFphhmNKmQQAWkAeB4KjMgUKgAFp7HMJZWj21BQbkcZ9DO0y2bmiHgE8EA4EkYoXLcoQRF2gBtABdXbMbXYAamMMmKZO5AADILeQU2XsplbGbJB3mCZ2IiDXNdMhW72Hz4T3KG9zHG0yNDLJQUAhpCEaxp0CQFvM5a1uM0yLpBY9VDeHRglK9h6GQMxgvOyE2tARIZHNmHkHhxGUdoNGMaxiLCLx7KxMa9SOMPUmGPJu2Tupx2bBpl3mZAVn2c5qiHyNn1beO+4rZtnuzeSZ26bX12CAff3kED32Hy9n3GxgUqRoxNqoFKgBZQRCOMfUmnxt3+DXcLN0mfAeB0LEGN6KB0lfvaJKH8v7umwn-f4O8HgQBYKrTgdAeQFGmodO2S1TKuXOsFGQQRvK+TwJNLgIBdpgAfA-aANZNLKmiAKHm-NBbBRsICNqMJPCqFRMgGAhddrHh0jIM6Y4wGy3csweIKQToG1IAfIy91TJwGQC8NmNIECik8BAHBPlFTDVGuNWg6AzhYjHCSPA1dkao3RgDEEtNRGfVIdKChpFlRBGatmT+1D+B8CAA)

Past issues, PRs, and discussions:
- 04/2018 https://github.com/DefinitelyTyped/DefinitelyTyped/pull/25385
- 11/2018 https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30820
- 08/2019 https://github.com/DefinitelyTyped/DefinitelyTyped/pull/37649
- 09/2019 https://github.com/DefinitelyTyped/DefinitelyTyped/issues/38342
- 10/2019 https://github.com/DefinitelyTyped/DefinitelyTyped/pull/38828
- 02/2020 https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42070
- 08/2021 https://github.com/DefinitelyTyped/DefinitelyTyped/pull/55010
- 08/2021 https://github.com/DefinitelyTyped/DefinitelyTyped/pull/55105
- 02/2022 https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58951
- 12/2022 https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63689
- 05/2023 https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/65579

@simonbuchan has responded to the majority of these.

**Reply by alex-bolenok-centralreach** [reply]
How about

```
type NewHandler<TEvent = any, TResult = any> =  
    | ((event: TEvent, context: Context, callback: Callback<TResult>) => void | undefined)
    | ((event: TEvent, context: Context) => Promise<TResult>)
    ;
```

It's a terrible hack (around a more terrible hack that the `void` type is), but it does invalidate `newHandler4`

**Reply by joeykilpatrick** [reply]
Still allows types that will fail at runtime:
```typescript
// New one incorrectly allows this. This will cause a runtime error when invoked by AWS Lambda.
const newHandler7: NewHandler = async (event, context, unexpected?: {prop1: {prop2: unknown}}) => { 
    if (unexpected && unexpected.prop1.prop2) {
        // ...
    }
    // ...
}
```

**Reply by alex-bolenok-centralreach** [reply]
What about this:

```ts
type SyncHandler<TEvent = any, TResult = any> = ((event: TEvent, context: Context, callback: Callback<TResult>) => void | undefined);
type AsyncHandler<TEvent = any, TResult = any> = ((event: TEvent, context: Context, callback: (arg: never) => void) => Promise<TResult>);

type NewHandler<TEvent = any, TResult = any> =  
    | SyncHandler<TEvent, TResult>
    | AsyncHandler<TEvent, TResult>
    ;

const handler1 = (event: {}, context: Context, callback: (value: number) => void) => { return; }
const handler2 = (event: {}, context: Context) => { return; }
const handler3 = (event: {}) => { return; }
const handler4 = async (event: {}, context: Context, callback: (value: number) => void) => { callback(1); return; } // Errors
const handler5 = async (event: {}, context: Context) => { return; }
const handler6 = async (event: {}) => { return; }
const handler7 = async (event: {}, context: Context, callback?: {prop1: number}) => { callback && callback.prop1; return; } // Errors
const handler8 = async (event: {}, context: Context, callback: (value: never) => void) => {  return; }

const newHandler1: NewHandler = handler1;
const newHandler2: NewHandler = handler2;
const newHandler3: NewHandler = handler3;
const newHandler4: NewHandler = handler4;
const newHandler5: NewHandler = handler5;
const newHandler6: NewHandler = handler6;
const newHandler7: NewHandler = handler7;
const newHandler8: NewHandler = handler8;
```

The way I understand it, we want to reflect the fact that the callback is there and it's a function (because AWS doesn't know what our function is going to return and sends the callback either way). Yet, we want to prevent the users from actually calling it if their type annotation says promise, because although it does work in the wild and the first one wins, it's technically undefined behavior.

One thing I don't like about this solution is that the union type kills type inference (i.e. typing a variable as `NewHandler` and assigning a lambda to it won't type the lambda parameters)

**Reply by simonbuchan** [reply]
> One thing I don't like about this solution is that the union type kills type inference (i.e. typing a variable as NewHandler and assigning a lambda to it won't type the lambda parameters)

Thing is: that's pretty much the entire point of the type, and the primary use of the package.

I simply don't understand the interest in not letting you call the callback if you're an async function. Are people actually hitting this repeatedly, and trying to figure out why the types didn't save them? Most of the time I see people either simply not understanding that you don't make the callback optional to not have it in your signature, or occasionally because they want to make their tests a little simpler (and not match the actual runtime behavior)

**Comment by darbio** [comment]
> One thing I don't like about this solution is that the union type kills type inference (i.e. typing a variable as NewHandler and assigning a lambda to it won't type the lambda parameters)

This kills this solution.

Does this really need "solving"? It is correct how it is (and we've discussed it for _years_ and always come back to the current implementation).

> Yet, we want to prevent the users from actually calling it if their type annotation says promise

Do we? That's not really the point of these definitions. Regardless, you don't _have_ to include the callback param, then you can't call it.

```
const handler: Handler<{ foo: string }, string> = async (event): Promise<string> => {
    return event.foo;
};
```

**Reply by alex-bolenok-centralreach** [reply]
> Does this really need "solving"?

It's not the cure for cancer, but it would make the world a little bit of a better place.

> Do we? That's not really the point of these definitions.

For me, one useful part of the definitions (among many others) is that they prevent you from doing things you're not supposed to do at compile time.

This looks like it *could* be one of those things, maybe deceptively so. Probably Typescript is not there yet and it's impossible to do that; or maybe there's a clever solution to this that no one has thought about yet.

I definitely agree that type inference is much more important that not allowing to call the callback in an async function. That's why it's a comment in a thread and not a PR.

**Reply by darbio** [reply]
> they prevent you from doing things you're not supposed to do at compile time.

The fact that, in a real life Lambda function, you can use either the callback OR the promise (and AWS will still allow you to do this in a first-in best-dressed fashion) makes me think that we should not stop it from being able to happen in these definitions.

> It's not the cure for cancer, but it would make the world a little bit of a better place.

I just don't get the real reason why we need to fix it? If you don't want to accidentally call the callback parameter, don't include it in your code:

```
const handler: Handler = async (event: APIGatewayProxyEvent): Promise< APIGatewayProxyResult > => {
    // no callback here!
    ...
};
```

Maybe I don't get the use case?

**Reply by alex-bolenok-centralreach** [reply]
> The fact that, in a real life Lambda function, you can use either the callback OR the promise (and AWS will still allow you to do this in a first-in best-dressed fashion) makes me think that we should not stop it from being able to happen in these definitions.

This behavior is not specified anywhere in the docs.

Some of us do possess this tribal knowledge that first-in, best-dressed is the way it works. Those who do learned it either from actually seeing it work this way, or from random people on the internet, not from AWS docs.

The AWS docs say use async / await if you can at all, but if you one of those callback weirdos, well, it's your funeral, we still let you use the callback. Nowhere in the docs I could find something definitely say what happens when you use both.

To me, it seems logical that if there is a simple, non-intrusive technical tool to prevent people from writing code that does both, people wouldn't mind to have it installed on their computers. This _looks_ like it might be a problem solvable by TypeScript.

Granted, this is not the biggest problem in the world of cloud computing. I'm still not opposed to investing a couple of hours into solving this particular problem, because I've been bitten by it.

> If you don't want to accidentally call the callback parameter, don't include it in your code

One thing I've seen happen back in the day when I was a consultant was something like this (simplified for clarity):

```ts
const oldLoggerFunctionFrom2016 = <TEvent extends any>(event: TEvent, context: Context) => {
  // log the event metric to statsd
};

const oldHandlerWrapperFrom2016 =
  <TEvent, TResult>(event: TEvent, context: Context, callback: Callback<TResult>, syncHandler: Handler<TEvent, TResult>): Handler<TEvent, TResult> =>
  (event, context, callback) => {
    syncHandler(event, context, callback);
    return oldLoggerFunctionFrom2016(event, context);
  };
```

which worked, until someone made `oldLoggerFunctionFrom2016` async and it started returning a promise. Then it mostly worked, because the handler was much faster than the logger, except when it wasn't. Took a while to figure it out.

It was not my code, and I'm not advocating writing code like that. Yet, this particular code was already there, and it would have been marked by TypeScript at compile time, had  something like what I'm suggesting been in place.

**Reply by simonbuchan** [reply]
I do agree in theory that it would be good (if minor) improvement, I just don't think it's worth the risk or churn of changing the central type of this package, or deprecating it and adding a replacement for such a small issue that's really more of a guideline than a clear bug. Note that the issue you described can only happen if the handler is using any as the response type - and that can cause far more serious issues that we can actually fix!

Typescript can't catch all your bugs, and mixing callbacks and async is always going to be a bug farm regardless of what AWS does.

**Reply by jamesorlakin** [reply]
I agree with the others here that this isn't a definition problem per se given an async function is provided the callback in practice anyway, so the types are correct. I can unfortunately only suggest not including it in your async handler's parameters.

> Nowhere in the docs I could find something definitely say what happens when you use both.

Aah - that's probably where the 'real' issue is: aspects of the Lambda runtime can be thinly documented. I've found this has been equally true for Java functions in the past too. While portions of the AWS docs are open source it seems [the section](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html#nodejs-handler-callback) around callback-based handlers isn't in the [linked repo](https://github.com/awsdocs/aws-lambda-developer-guide/tree/main/doc_source/nodejs-handler.md). I was going to suggest submitting changes there to clarify async handlers still receive the callback but it seems this isn't an accessible avenue.

> The AWS docs say use async / await if you can at all, but if you one of those callback weirdos, well, it's your funeral

But on the other hand, it's 2023. `async/await` has been about for over half a decade at this point and the last non-async Node.js runtime version (pre v8.x) was fully deprecated on [Mar 5, 2020](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html#runtime-support-policy), not to forget TypeScript users will be transpiling anyway. Is it even a consideration to write *new* callback-based handlers anymore?

**Reply by simonbuchan** [reply]
There's still a few callback based libraries out there that are a little easier to use in small functions with a callback. And a few weirdos who just don't like promises.

Of course, if we were to unilaterally deprecate callbacks, even by bumping major, I'm quite sure we'd get complaints (and rightly so, if we were causing them to have to update and possibly break existing code to continue getting updates)

Maybe petition AWS to remove the callback 🤷

**Reply by darbio** [reply]
> Maybe petition AWS to remove the callback 🤷

Probably the sanest option, however, I am sure that there are still some lambda's out there which haven't been updated and still use callbacks. (I suspect that I have one somewhere from yesteryear, before asynchrony, doing some trivial task which I would miss if it didn't work but don't remember about right now!)

> Of course, if we were to unilaterally deprecate callbacks, even by bumping major

If AWS do actually ever deprecate callbacks, then that is when we should create a new handler definition for the new Lambda architecture, leaving the old one in place until such a time that it is non-deployable anymore, after which it can be removed.

**Reply by alex-bolenok-centralreach** [reply]
How about we split it into two signatures, leave the name `Handler` for the async one and rename the old one to `SyncHandler` or anything?

**Reply by simonbuchan** [reply]
And also update every trigger handler definition (more than one in some triggers!) in `types/aws-lambda/trigger/`, e.g.

```patch
---types/aws-lambda/trigger/dynamodb-stream.d.ts
+++types/aws-lambda/trigger/dynamodb-stream.d.ts
-export type DynamoDBStreamHandler = Handler<DynamoDBStreamEvent, DynamoDBBatchResponse | void>;
+export type DynamoDBStreamSyncHandler = SyncHandler<DynamoDBStreamEvent, DynamoDBBatchResponse | void>;
+export type DynamoDBStreamHandler = Handler<DynamoDBStreamEvent, DynamoDBBatchResponse | void>;
```

and also we're changing the meaning of the definition, so bump the major version, which breaks the link to the runtime version...

---

I'd say it's ok to do that with a non-breaking update, e.g. just adding `AsyncHandler` without changing the existing `Handler`, but it's a lot of work that most likely wouldn't be noticed by most users?

**Reply by alex-bolenok-centralreach** [reply]
> which breaks the link to the runtime version

If I'm reading this right, the package version is 8.10 now, which seems to correspond to the first node version to support promises.

`nodejs8.x` as a lambda runtime environment stopped being supported three years ago.

That said, I'm not even sure why we should link to the version of the runtime environment, rather than the "version" of the handler API (whatever it might mean). New node releases do not necessarily introduce type changes on their own, and the handle API (theoretically) can change even within the same runtime version (of course that's highly unlikely)

The closest thing to a new "release" of the lambda handler API seems to be a change on the [handler documentation page](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html). If you look through the history of these updates, they go from "you might also want to try this cool new async thing" to "async and callback both work" to "use async if you can at all, and here's the callback for you in the cheap seats".

This last version is a closest thing to the deprecation notice. Per semver, we should have bumped the minor version and mark callback as deprecated when it had happened.

**Reply by simonbuchan** [reply]
I've actually wanted to see this package split out into a public API with a package for the handler and a separate package for each trigger type - this would make breaking changes for each package be a lot less disruptive for all the people taking dependencies on this package.

**Reply by darbio** [reply]
> This last version is a closest thing to the deprecation notice.

Until AWS give explicit notice that they are deprecating the callback functionality, we won't be deprecating it from the types or marking it as such. 

Our job here is not to be opinionated and make people do things a specific way.

Our job is to provide typings for AWS Lambda which aid people to code for it using Typescript in the manner in which they want to use it.

**Reply by ethanherbertson** [reply]
Not sure if this is the right place to bring this up, but AWS has now given explicit notice that they are deprecating the callback functionality: https://github.com/aws/aws-lambda-nodejs-runtime-interface-client/issues/137

**Reply by darbio** [reply]
> AWS has now given explicit notice that they are deprecating the callback functionality: [aws/aws-lambda-nodejs-runtime-interface-client#137](https://github.com/aws/aws-lambda-nodejs-runtime-interface-client/issues/137)

From https://github.com/aws/aws-lambda-nodejs-runtime-interface-client/issues/137

> This proposal applies to Node.js 24 and later runtimes only. We will continue to support both callback and async/await function handler signatures in our existing Node.js 18, Node.js 20, and Node.js 22 runtimes.

We need to remain backwards compatible as there will be people still using older versions of Node / aws lambda runtime environments. This has the potential to be a massive pain in the backside..

Should we change versions to follow Node version, rather than lambda runtime version? (e.g. `@types/aws-lambda@24.0.0` where 24 is the Node version you are using)

It doesn't seem right.. but it could solve backwards compatibility issues.

**Reply by ethanherbertson** [reply]
Version pinning would allow users on deprecated runtimes to continue to use the old type(s). I don't see a reason not to have this just be included in the next major version of the module?

**Comment by joeykilpatrick** [comment]
Since we're all back to discussing this, I want to bring up an idea that was raised in [this PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63689#issuecomment-1362101659) that showed some promise. The PR was abandoned, but I think the idea is very clever. Would appreciate some feedback from @simonbuchan and others.

The idea would be to change the `Handler` definition to
```typescript
type AsyncHandler<TEvent = unknown, TResult = unknown> =
  ((event: TEvent, context: Context, callback: Callback<TResult>) => Promise<TResult>) &
  ((event: TEvent, context: Context) => Promise<TResult>);

export type Handler<TEvent = unknown, TResult = unknown> =  
    | ((event: TEvent, context: Context, callback: Callback<TResult>) => void)
    | AsyncHandler<TEvent, TResult>
    ;
```

This would prevent handlers that both return a promise *and* use the callback argument, which is apparently what the people want.

It still does not allow users to (incorrectly) call their own handlers without providing a callback argument. We will always get PRs and issues trying to "fix" this one.

There were concerns about if this would be usable with earlier version Typescript (< 4.3). If we found a workable solution here, I think it would make a lot of people happy.
